components/recurring-contributions/UpdatePaymentMethodPopUp.js

```diff
@@ -0,0 +1,280 @@
+  const intl = useIntl();
+  const [selectedPaymentMethod, setSelectedPaymentMethod] = useState(null);
+  const [submitUpdatePaymentMethod, { loadingUpdatePaymentMethod }] = useMutation(updatePaymentMethodMutation, {
+    context: API_V2_CONTEXT,
+  });
+
+  // data handling\
+  const minBalance = 50; // Minimum usable balance for virtual card
+
+  const getPaymentMethodIcon = pm => {
+    if (pm.type === 'creditcard') {
+      return <CreditCard />;
+    } else if (pm.type === 'virtualcard') {
+      return <GiftCard />;
+    } else if (pm.type === 'prepaid') {
+      return <MoneyCheck width="26px" height="18px" />;
+    }
+  };
+
+  const getPaymentMethodMetadata = pm => {
+    if (pm.type === 'creditcard') {
+      const expiryDate = paymentMethodExpiration(pm);
+      return (
+        <FormattedMessage
+          id="ContributePayment.expiresOn"
+          defaultMessage="Expires on {expiryDate}"
+          values={{ expiryDate }}
+        />
+      );
+    } else if (pm.type === 'virtualcard') {
+      if (pm.balance < minBalance) {
+        return (
+          <FormattedMessage
+            id="ContributePayment.unusableBalance"
+            defaultMessage="{balance} left, balance less than {minBalance} cannot be used."
+            values={{
+              balance: formatCurrency(pm.balance, pm.currency),
+              minBalance: formatCurrency(minBalance, pm.currency),
+            }}
+          />
+        );
+      } else if (pm.expiryDate) {
+        return (
+          <FormattedMessage
+            id="RecurringContributions.balanceAndExpiry"
+            defaultMessage="{balance} left, expires {expiryDate}"
+            values={{
+              expiryDate: <FormattedDate value={pm.expiryDate} month="numeric" year="numeric" />,
+              balance: formatCurrency(pm.balance, pm.currency),
+            }}
+          />
+        );
+      } else {
+        return (
+          <FormattedMessage
+            id="ContributePayment.balanceLeft"
+            defaultMessage="{balance} left"
+            values={{ balance: formatCurrency(pm.balance, pm.currency) }}
+          />
+        );
+      }
+    } else if (['prepaid', 'collective'].includes(pm.type)) {
+      return (
+        <FormattedMessage
+          id="ContributePayment.balanceLeft"
+          defaultMessage="{balance} left"
+          values={{ balance: formatCurrency(pm.balance, pm.currency) }}
+        />
+      );
+    }
+  };
+
+  const generatePaymentsOptions = () => {
+    // Add collective payment methods
+    const paymentMethodsOptions = get(data, 'Collective.paymentMethods', [])
+      .filter(pm => !(pm.service === 'paypal' && pm.type === 'adaptive'))
+      .map(pm => ({
+        key: `pm-${pm.id}`,
+        title: getPaymentMethodName(pm),
+        subtitle: getPaymentMethodMetadata(pm),
+        icon: getPaymentMethodIcon(pm),
+        paymentMethod: pm,
+        disabled: pm.balance < minBalance,
+        id: pm.id,
+      }));
+
```

Comment:
I would like to set the default checked payment method in the radio list to the one that is currently attached to the order. That will require a bit of extra Graphql work but for now I was literally just trying to set it as `setSelectedPaymentMethod(paymentMethodOptions[0])` (see line 56) but this creates an infinite render loop. I looked into `useEffect` and tried writing a custom hook but I unfortunately haven't written/used them before, and remain a bit unsure about the implementation. I tried putting it in this function and the React linter said that was not allowed (and it didn't seem to work anyways). So how could I use `useEffect` here to set a default checked radio option when I create the list of payment methods?

Reply:
In a nutshell, `useEffect` is called on mount or when one of the items of the array passed as the second argument changes. So this:

```jsx
useEffect(() => {
  setValue(someArg);
}, [someArg])
```

will set the value to `someArg` on mount, and each time the value changes. You can see an usage example here: https://github.com/opencollective/opencollective-frontend/blob/379a1f864f0d0b4bd54c05f8cceee719791cd1b5/components/expenses/ExpenseForm.js#L187-L191